Désinscription

Alex-D opened the pull request:
| Q | R |
| --- | --- |
| Correction de bugs ? | Oui |
| Nouvelle Fonctionnalité ? | Oui |
| Tickets concernés | #818 |

Cette PR est la suite de la #1400 initiée par @artragis mais qui est directement sur le dépôt principal permettant ainsi à chacun d'y contribuer plus directement au vu de la complexité de celle-ci.

Elle également là suite à une erreur de ma part alors que j'ai souhaité corrigé un template sans lien avec cette PR.

Note : cette PR est basée sur `dev` et non plus sur `master` comme l'était la PR #1400.

En ayant une branche sur le dépôt principal, il faut donc faire des PR sur cette branche qui sera elle-même mergée via cette présente PR. Procéder ainsi permet de valider étape par étape les diverses modifications plutôt que de devoir revalider toutes les modifications. Aussi, chacun peu contribuer s'il le souhaite.
# QA
## Interface utilisateur (front)
- Lien de désinscription accessible via **Paramètres** (en en haut à droite de toute les pages sur l'avatar / roue dentée) puis **Désinscription** dans la barre latérale ;
- Le lien mène alors vers une page expliquant les conséquences de sa désinscription avec un bouton rouge en bas de celle-ci ;
- Le clic sur le bouton rouge ouvre une boite modale qui constitue le dernier avertissement avant le déclenchement du processus de désinscription.
## Effets de la désinscription (back)
- Suppression du profil, libèrant le pseudo et l'adresse courriel pour les futures inscriptions ;
- Le membre est déconnecté ;
- Les données du membre sont anonymisées : 
  - messages du forum ayant comme pseudo "Anonyme" ;
  - les sujets du forum restent ouverts mais ont comme auteur "Anonyme" ;
  - les galeries non liées à un tutoriel sont données à "Auteur externe" (puisque l'image peut être considérée comme venant d'un "auteur") avec droit de lecture et d'écriture ;
  - les MP sont anonymisés et le membre quitte les discussions auxquelles il participait ;
  - les commentaires de tutoriels/articles ont comme pseudo "Anonyme" ;
  - les articles et tutoriels suivent ces règles : 
    - si le tutoriel/article a été écrit par **plusieurs personne**; le membre est retiré de la liste des auteurs ;
    - si le tutoriel/article est **publié***, il passe sur le compte "Auteur externe", une demande expresse sera nécessaire au retrait complet de ses contenus ;
    - si le tutoriel/article n'est pas publié (brouillon, bêta, validation) il est supprimé ainsi que la galerie associée.
# Historique

Pour lire/voir/comprendre les diverses évolutions, il est possible que la réponse se trouve dans la PR #1400, passer rapidement dessus (Ctrl+F est votre ami :) pourra peut-être vous apporter des réponses.

artragis committed:
Permet la désinscription côté back

artragis committed:
Fixture et corrections

artragis committed:
Anonym. galeries et MP + suppr tutos en redacion

artragis committed:
Problème de suppression des tutos

artragis committed:
Ajout de test et debug

artragis committed:
Test des forum pour desinscription

artragis committed:
Intégration au front + typo

artragis committed:
utilise la requête en POST

artragis committed:
Intégration au front sans modale

artragis committed:
Bouge btn de desinscript + tests articles et tutos

artragis committed:
Ne change l'éditeur que s'il y a eu edition

artragis committed:
Nettoyage des tests

artragis committed:
Désinscrit participants d'une discussion privée

artragis committed:
Corrige les galeries pour désinscription

artragis committed:
Page d'avertissement + fixtures

artragis committed:
Documentation des fixtures

artragis committed:
Suppresion des MP désinscrits

Alex-D committed:
Amélioration du front de la désinscription

artragis committed:
Suppr tuto complet lorsque le membre se désinscrit

artragis committed:
Suppr tuto complet lorsque le membre se désinscrit

artragis committed:
Override des settings dans le test

artragis committed:
Test unitairement la suppression des MP

artragis committed:
Nettoie le dossier des tutos lors du test

artragis committed:
Vérifie que les dossiers existent dans le tearDown

artragis committed:
Fix manquements fonctionnels sur MP et galeries

- Pour les galeries des tutos non publiés, elles doivent être supprimées. En effet, plus personne n'y sera rattaché (même pas auteur externe) et leurs tutos respectifs sont viré du disque dur. Du coup il faut les virer aussi pour éviter les données fantôme.

- Pour les MP. Prenons le cas d'un MP a trois personnes, A, B et C. Ils s’échangent tous plein de MP dans un fil à tout les trois. Si B quitte le site, A et C doivent pouvoir continuer le MP sans le perdre. Donc un membre partant doit juste quitter le MP sans qu'il soit supprimé.

artragis committed:
Rétablit la modale

firm1 commented:
Etant donné que cette PR est ouverte après le merge de la [PR de doc](https://github.com/zestedesavoir/zds-site/pull/1409) on a plus d'excuse. Comme on l'a dit lors du zest'meeting, Chaque nouvelle feature doit être rajoutée dans la doc pour être éligible au merge.

Donc cette PR devrait alimenter la doc

artragis commented:
je veux bien l'écrire mais comme cette PR n'est plus sur mon dépôt, je dois faire une PR sur desinscription pour que ça marche, c'est ça?

firm1 commented:
tout à fait

artragis committed:
Corrige le bug de supression des MP

     et ajout de la documentation

artragis committed:
Met à jour le comportement pour les topics privés

       (changement d'auteur)

artragis committed:
Ajout de documentation sur les utilisateurs

artragis committed:
Evite doublon utilisateur anonymisation des MP

Alex-D commented:
Je pose un lock sur cette PR : je tente un rebase pour la rendre compatible avec les modifs pep8 faites récemment.

Alex-D committed:
Validation Pep8

Alex-D committed:
Amelioratron de la doc membre pour la desinscription

fix images

typo

ajout précision

Amélioration du contenu

Ajout images dans le depot et typo

Alex-D commented:
Voilà, unlocked :)

pierre-24 commented:
Rapport de QA : 

Selon la recette donnée par @Eskimon sur IRC, j'ai créé, à l'aide d'un user U,
- Un MP dont le destinatiaire était admin, un MP dont le destinataire était admin et staff
- Un topic sur le forum, j'ai commenté un topic de admin, j'ai éditer le message
- Comme seul auteur, un tutoriel en draft, en béta, en validation et en ligne ainsi qu'un article en draft, en validation et en ligne.
- Comme second auteur avec admin, la même chose.

Je suis repassé dessus avec admin pour,
- Créer les articles et tutos comme précisé plus hat
- Écrire un MP impliquant seulement U, mais aussi un MP avec U et staff
- Créer un sujet, ou U commente, comme précisé plus haut
- Éditer un message de U

... Et puis j'ai désinscrit U.

**Résultats :**
- article et tuto en ligne avec U pour seul auteur : changé en auteur externe, OK
- article et tuto avec U en second auteur : retiré de la liste, OK
- Dans le forum, U remplacé par anonymous partout, même dans l'édition de son propre message, OK
- MP de U :tout en anonymous, OK
  - avec admin : admin seul, OK
  - avec admin et staff : U plus dans la liste, OK
- MP de admin a U : 
  - avec U seul : admin reste seul, OK
  - avec U et staff : U plus dans la liste, OK
- Tuto et articles de U en draft, béta et validation : disparu de la BDD, OK. À noter que les objets "Validation" ne sont eux pas supprimés de la BDD si le tuto était déjà réservé : **éventuellement pas OK** (peut poser des problèmes par la suite)
- Tuto et articles de admin avec U : U a disparu de la liste, OK

Test de réinscription avec le même pseudo, OK. À noter que les _tokens d'inscriptions_ ne disparaissent pas de la base de donnée, **éventuellement pas OK** (peut poser problème par la suite). 

Bref, je crois qu'on tient le truc, là. J'ai encore une PR de modif de doc' à faire,  si jamais, donc mergez pas tout de suite.

artragis commented:
OK, je crois qu'il faut créer un événement de PostDelete dans le modèle
des tutoriels qui supprime le token de validation comme je le fais avec
les galeries.
Le 09/09/2014 20:32, Pierre Beaujean a écrit :

> Rapport de QA :
> 
> Selon la recette donnée par @Eskimon https://github.com/Eskimon sur
> IRC, j'ai créé, à l'aide d'un user U,
> - Un MP dont le destinatiaire était admin, un MP dont le
>   destinataire était admin et staff
> - Un topic sur le forum, j'ai commenté un topic de admin, j'ai
>   éditer le message
> - Comme seul auteur, un tutoriel en draft, en béta, en validation et
>   en ligne ainsi qu'un article en draft, en validation et en ligne.
> - Comme second auteur avec admin, la même chose.
> 
> Je suis repassé dessus avec admin pour,
> - Créer les articles et tutos comme précisé plus hat
> - Écrire un MP impliquant seulement U, mais aussi un MP avec U et staff
> - Créer un sujet, ou U commente, comme précisé plus haut
> - Éditer un message de U
> 
> ... Et puis j'ai désinscrit U.
> 
> _Résultats :_
> - article et tuto en ligne avec U pour seul auteur : changé en
>   auteur externe, OK
> - article et tuto avec U en second auteur : retiré de la liste, OK
> - Dans le forum, U remplacé par anonymous partout, même dans
>   l'édition de son propre message, OK
> - MP de U :tout en anonymous, OK
>     o avec admin : admin seul, OK
>     o avec admin et staff : U plus dans la liste, OK
> - MP de admin a U :
>     o avec U seul : admin reste seul, OK
>     o avec U et staff : U plus dans la liste, OK
> - Tuto et articles de U en draft, béta et validation : disparu de la
>   BDD, OK. À noter que les objets "Validation" ne sont eux pas
>   supprimés de la BDD si le tuto était déjà réservé :
>   _éventuellement pas OK_ (peu poser des problèmes par la suite)
> - Tuto et articles de admin avec U : U a disparu de la liste, OK
> 
> Test de réinscription avec le même pseudo, OK. À noter que les /tokens
> d'inscriptions/ ne disparaissent pas de la base de donnée,
> _éventuellement pas OK_ (peut poser problème par la suite).
> 
> Bref, je crois qu'on tient le truc, là. J'ai encore une PR de modif de
> doc' à faire, de toute façon.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1469#issuecomment-55013196.

pierre-24 commented:
Bien vu, j'ai oublié les galeries : 
- Galerie créée par U : disparues de la BDD, OK
- Galeries en commun avec admin : U à disparu de la liste, OK

artragis commented:
Pour ces deux points : 

> - Test de réinscription avec le même pseudo, OK. À noter que les tokens d'inscriptions ne disparaissent pas de la base de donnée, éventuellement pas OK (peut poser problème par la suite). 
> - Tuto et articles de U en draft, béta et validation : disparu de la BDD, OK. À noter que les objets "Validation" ne sont eux pas supprimés de la BDD si le tuto était déjà réservé : éventuellement pas OK (peut poser des problèmes par la suite)

le meilleur moyen serait de changer le modèle de données de base pour forcer le delete en cascade.
Ce qui signifierait l'écriture d'une migration.

firm1 commented on zds/member/tests/tests_views.py:
```diff
@@ -202,3 +446,18 @@ def test_nonascii(self):
+            rmtree(settings.REPO_PATH)
+        if os.path.isdir(settings.REPO_PATH_PROD):
+            rmtree(settings.REPO_PATH_PROD)
+        if os.path.isdir(settings.REPO_ARTICLE_PATH):
```
problème de doublon ici le code supprime `settings.REPO_ARTICLE_PATH` et `settings.MEDIA_ROOT` deux fois

firm1 replied:
Ce problème de doublon n'a pas été réglé

firm1 commented on zds/tutorial/models.py:
```diff
@@ -333,6 +334,14 @@ def get_conclusion_online(self):
 
             return conclu_contenu.decode('utf-8')
 
+    def delete_entity_and_tree(self, using=None):
```
a quoi sert le paramètre using ici ? il n'est même pas utilisé dans la fonction

artragis replied:
un manque de maîtrise de django au départ, j'avais tenté d'overrider la
fonction "delete" et je me suis un peu trompé.

Désolé de ne pas avoir corrigé avant.
Le 11/09/2014 14:25, firm1 a écrit :

> a quoi sert le paramètre using ici ? il n'est même pas utilisé dans la
> fonction

firm1 commented on zds/article/models.py:
```diff
@@ -90,6 +92,12 @@ class Meta:
     def __unicode__(self):
         return self.title
 
+    def delete_entity_and_tree(self, using=None):
```
pourquoi ne pas avoir mis un `@receiver(post_delete)` ici ? pour surcharger la fonction de delete tout simplement ?

artragis replied:
jocker
Le 11/09/2014 14:29, firm1 a écrit :

> In zds/article/models.py:
> 
> > @@ -90,6 +92,12 @@ class Meta:
> >      def **unicode**(self):
> >          return self.title
> > -    def delete_entity_and_tree(self, using=None):
> 
> pourquoi ne pas avoir mis un |@receiver(post_delete)| ici ? pour
> surcharger la fonction de delete tout simplement ?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1469/files#r17417722.

firm1 commented on zds/member/views.py:
```diff
@@ -85,6 +89,82 @@ def index(request):
+    external = get_object_or_404(User, username=EXTERNAL_USER)
+    current = request.user
+    for tuto in request.user.profile.get_tutos():
+        # we delete article only if not published with only one author
```
> `# we delete article only if not published with only one author`

à remplacer par
`# we delete tutorial only if not published with only one author`

firm1 replied:
cette remarque n'a pas été prise en compte

firm1 replied:
remarque toujours pas prise en compte :(

SpaceFox replied:
Fais donc une PR :)

firm1 commented on zds/member/views.py:
```diff
@@ -85,6 +89,82 @@ def index(request):
+    # all messages anonymisation (forum, article and tutorial posts)
+    for message in Comment.objects.filter(author=current):
+        message.author = anonymous
+        if message.editor is not None and message.editor.username == current.username:
```
> ``` python
> if message.editor is not None and message.editor.username == current.username:
>     message.editor = anonymous
> ```
> 
> Cette partie du code est inutile, car elle vient en doublon avec ton code du dessous
> 
> ``` python
> for message in Comment.objects.filter(editor=current):
>     message.editor = anonymous
>     message.save()
> ```

artragis replied:
C'est la partie en dessous qui est inutile. Ce qui est étrange c'est que
je suis sûr de l'avoir enlevé ce code.
Le 11/09/2014 14:40, firm1 a écrit :

> In zds/member/views.py:
> 
> > -                tuto.authors.add(external)
> > -            tuto.authors.remove(current)
> > -            tuto.save()
> > -    for article in request.user.profile.get_articles():
> > -        # we delete article only if not published with only one author
> > -        if not article.on_line() and article.authors.count() == 1:
> > -            article.delete_entity_and_tree()
> > -        else:
> > -            if article.authors.count() == 1:
> > -                article.authors.add(external)
> > -            article.authors.remove(current)
> > -            article.save()
> > -    # all messages anonymisation (forum, article and tutorial posts)
> > -    for message in Comment.objects.filter(author=current):
> > -        message.author = anonymous
> > -        if message.editor is not None and message.editor.username == current.username:
> 
> ```
> if message.editor is not None and message.editor.username == current.username:
>     message.editor = anonymous
> 
> Cette partie du code est inutile, car elle vient en doublon avec
> ton code du dessous
> 
> for message in Comment.objects.filter(editor=current):
>     message.editor = anonymous
>     message.save()
> ```
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1469/files#r17418099.

firm1 replied:
la partie en dessous englobe la partie du dessus. Car avec uniquement le morceau du dessus, si un valido avait modifié un message d'un membre, et que ce valido se désinscrit, le message sera toujours considéré comme édité par le valido parti

firm1 commented on zds/member/views.py:
```diff
@@ -85,6 +89,82 @@ def index(request):
+            anonymousGallery.gallery = gallery.gallery
+            anonymousGallery.save()
+        else:
+            gallery.delete()
```
avec ceci, on ne risque pas de supprimer des galleries pour une tuto publique ?

artragis replied:
faut que je vérifie, mais il me semble qu'au moment où je mets external
il y a un truc qui est fait.
Le 11/09/2014 14:57, firm1 a écrit :

> In zds/member/views.py:
> 
> > -    for topic in PrivateTopic.objects.filter(participants__in=[current]):
> > -        topic.participants.remove(current)
> > -        topic.save()
> > -    for topic in Topic.objects.filter(author=current):
> > -        topic.author = anonymous
> > -        topic.save()
> > -    TopicFollowed.objects.filter(user=current).delete()
> > -    for gallery in UserGallery.objects.filter(user=current):
> > -        if gallery.gallery.get_users().count() == 1:
> > -            anonymousGallery = UserGallery()
> > -            anonymousGallery.user = external
> > -            anonymousGallery.mode = "w"
> > -            anonymousGallery.gallery = gallery.gallery
> > -            anonymousGallery.save()
> > -        else:
> > -            gallery.delete()
> 
> avec ceci, on ne risque pas de supprimer des galleries pour une tuto
> publique ?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1469/files#r17418760.

firm1 replied:
De toute façon il faudrait sortir l'instruction du `else` car même si on passe la gallery à external, on doit aussi virer les droits du désinscrit dessus

artragis replied:
Très clairement, la condition actuelle est OK car elle s'appuie sur les supressions qui ont déjà eu lieu avant.
Afin de rendre le code plus maintenable, je vais ajouter un commentaire, mais actuellement il n'y a rien à changer dans ce code.

firm1 commented:
Petit détail a vérifier aussi par la QA : lors de la désinscription, pensons à vérifier que les alertes partent vraiment.

dralliw committed:
Merge pull request #1491 from pierre-24/doc_desinscription

[PR désinscription] Amélioration de la doc membre

firm1 commented:
Pour infor je viens de compiler la doc relative a cette branche, et on peut l'apercevoir ici : http://zds-site.readthedocs.org/fr/desinscription/member/member.html

SpaceFox commented:
Le pseudo du membre anonyme est "Anonyme" et pas "anonymous".

firm1 commented:
> Le pseudo du membre anonyme est "Anonyme" et pas "anonymous".

Etant donné que le pseudo du membre est un paramètre du settings.py la doc est cohérente avec le code.
Je ne suis pas sur que ce soit une bonne idée (pour l'aspect opensource) que le settings.py soit lié au site zestedesavoir.com, c'est plutôt le rôle du settings_prod.py de s'adapter aux spécificités du site.

SpaceFox commented:
Du coup la langue utilisée est incohérente avec les autres pseudos, et ça
sent la MEP foireuse d'ici.
Le 12 sept. 2014 11:59, "firm1" notifications@github.com a écrit :

> Le pseudo du membre anonyme est "Anonyme" et pas "anonymous".
> 
> Etant donné que le pseudo du membre est un paramètre du settings.py la doc
> est cohérente avec le code.
> Je ne suis pas sur que ce soit une bonne idée (pour l'aspect opensource)
> que le settings.py soit lié au site zestedesavoir.com, c'est plutôt le
> rôle du settings_prod.py de s'adapter aux spécificités du site.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1469#issuecomment-55382972
> .

pierre-24 commented:
> Du coup la langue utilisée est incohérente avec les autres pseudos, et ça
> sent la MEP foireuse d'ici.

Je comprend pas. Les pseudos, dans le code, sont tous en anglais (les nom d'utilisateurs ne le sont pas, là je suis d'accord), mais je ne vois pas ou ça pose problème à un moment ou un autre ? il suffit de changer le `setting_prod` comme y faut, nan ? o_O

artragis committed:
permet de nettoyer complètement les objets attachés à l'utilisateur

artragis committed:
pep8

Eskimon commented:
Rappel : L’évolution du site est complètement figée tant qu'on avance pas ici (et notamment sur cette PR : https://github.com/zestedesavoir/zds-site/issues/1507 qui est censé corriger des points bloquants)

SpaceFox commented:
Hmmm, on a des conflits ici.

Quelqu'un a des nouvelles ?

dralliw committed:
Merge pull request #1507 from artragis/desinscription_cleanup

[Désinscription] Permet de nettoyer complètement les objets attachés à l'utilisateur

pierre-24 commented:
Bon, je viens de faire de la QA vite fait bien fait (de la #1507, en fait, mais vous avez été trop vite), j'ai créé un mini-tuto et un article : 
- Les tokens de validations disparaissent bien, OK
- Le token d'inscription disparait aussi, donc OK aussi. 

Donc ...À MERGER DÈS QUE LES CONFLITS SONT RÉGLÉS !!! (ça fait des jours que je rêve d'écrire ça sur cette PR)

pierre-24 committed:
Merge remote-tracking branch 'upstream/dev' into fix-desinsc-conflicts

pierre-24 committed:
Rectification d'un test

pierre-24 committed:
Merge remote-tracking branch 'upstream/dev' into fix-desinsc-conflicts

Eskimon committed:
Merge pull request #1513 from pierre-24/fix-desinsc-conflicts

Fix des conflits de désinscription

Eskimon commented:
Ping @SpaceFox @ArnaudCalmettes @pierre-24 @artragis @firm1 @gustavi et les autres... Il est temps de se concentrer une dernière fois... Cette PR de désinscription est considéré comme **FINI** d'après certains d'entres nous. Cependant, avant le merge (qui va péter plusieurs PR) il serait bon de s'assurer que l'on oublie rien. Merci à tous de checkup, de poser vos questions, de lever tout les doutes et de balancer vos +1 si OK pour merge...

Une fois ce merge fait, direction la QA grandeur nature en préprod et en avant pour la V1.1 et son lot de bonheur :)

pierre-24 commented:
+1 pour le merge, pour moi c'est bon ;)

Alex-D commented:
Est-ce qu'on ferais pas un merge/rebase de dev dans desinscription (ou une autre branche) pour avoir une branche testable telle qu'après le merge, le tout en préprod pour tout tester et vérifier que ça ne casse rien ?

firm1 commented on README.md:
```diff
@@ -49,10 +49,12 @@ python manage.py loaddata fixtures/*.yaml
 Cela va créer plusieurs entitées :
 
-* 3 utilisateurs (utilisateur/mot de passe) :
+* 5 utilisateurs (utilisateur/mot de passe) :
```
A corriger, avec l'ajout des deux utilisateurs on aura plus 5 mais 6 utilisateurs dans les fixtures

artragis replied:
l'utilisateur itrema doit-il être documenté? En effet, il semble n'être là que pour mettre un autre membre en cas de besoin d'un membre vide pour les tests, mais je ne suis pas sûr qu'il soit vraiment utile.

pierre-24 replied:
Si il est dans les fixtures, il doit être documenté (preuve en est, je pensais de bonne fois que c'est moi qui l'avait créé).

artragis replied:
Ok, du coup je passe mon tour, parce que moi et les mots de passe dans
les fixtures, ça fait deux.
Il faut donc que quelqu'un corrige cette faute et celle de mon
commentaire ligne 108 du fichier `zds/member/views.py`.
Le 17/09/2014 08:41, Pierre Beaujean a écrit :

> In README.md:
> 
> > @@ -49,10 +49,12 @@ python manage.py loaddata fixtures/*.yaml
> > 
> >  Cela va créer plusieurs entitées :
> > 
> > -\* 3 utilisateurs (utilisateur/mot de passe) :
> > +\* 5 utilisateurs (utilisateur/mot de passe) :
> 
> Si il est dans les fixtures, il doit être documenté (preuve en est, je
> pensais de bonne fois que c'est moi qui l'avait créé).
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1469/files#r17648060.

pierre-24 replied:
... Je ne connais pas non plus le mot de passe de cet utilisateur. Voyons voir qui l'as rajouté dans les fixtures ... Arf, c'est "ïtrema/ïtrema" :p

pierre-24 replied:
Par contre, @artragis, pour "Auteur externe", en local, j'arrive pas à le connecter, donc t'es sur que le mot de passe est bien "external" ?

firm1 commented on doc/sphinx/source/member/member.rst:
```diff
@@ -1,7 +1,105 @@
+- admin/admin : Utilisateur avec les droits d'un staff et d'un admin
+- anonymous/anonymous : Utilisateur qui permet l'anonymisation des messages sur les forums, dans les commentaires d'articles et de tutoriels ainsi que dans les MPs
+- Auteur externe/external : Utilisateur qui permet de récupérer les tutoriels d'anciens membres et/ou de publier des tutoriels externes.
+
```
oublie de l'user itrema dans les fixtures

pierre-24 replied:
_My bad_, je croyais que cet utilisateur faisait partie de ceux que j'ai généré au fil de mes tests.

firm1 commented on zds/member/views.py:
```diff
@@ -85,6 +89,89 @@ def index(request):
+            anonymousGallery.mode = "w"
+            anonymousGallery.gallery = gallery.gallery
+            anonymousGallery.save()
+        gallery.delete()
```
Petit souci avec ce code. Tel qu'il est écrit, ça signifique que si j'ai un user `A` avec les droits d'écritures, et un user `B` avec les droits de lectures sur la gallerie, si l'user `A` décide de partir, il ne restera plus qu'un user `B` dans la galerie et ce dernier n'aura plus les droits d'écriture sur cette gallerie.

Il faut rajouter external dans la gallerie, s'il n'y a plus personne qui a les droits d'écriture sur cette gallerie

pierre-24 replied:
Vrai question : qu'est ce que ça change ? Dans les faits, personne n'as accès à "Auteur externe", donc qu'est ce que ça change d'avoir une gallerie en lecture seule ?

artragis replied:
En quoi cela pose-t-il problème?

Le principe (et je pense que le commentaire est assez claire, non?)
c'est que ce ne sont que des galeries sur le _forum_. On ne les garde
que pour éviter d'avoir des 404 sur les messages du forum qui ont été
anonymisés.

A priori, sauf si la gallerie a été partagée en écriture, il n'y a aucun
besoin à ce que quelqu'un récupère l'écriture puisque le but de garder
cette galerie est justement d'éviter des problèmes de _lecture_ uniquement.
Le 17/09/2014 00:12, firm1 a écrit :

> In zds/member/views.py:
> 
> > -    # - unpublished articles with only the unregistering member as an author
> > -    # - all category associated with those entites (have a look on article.delete_entity_and_tree
> > -    #   and tutorial.delete_entity_and_tree
> > -    # So concerning galleries, we just have for us
> > -    # - "personnal galleries" with only one owner (unregistering user)
> > -    # - "personnal galleries" with more than one owner
> > -    # so we will just delete the unretistering user ownership and give it to anonymous in the only case
> > -    # he was alone so that gallery is not lost
> > -    for gallery in UserGallery.objects.filter(user=current):
> > -        if gallery.gallery.get_users().count() == 1:
> > -            anonymousGallery = UserGallery()
> > -            anonymousGallery.user = external
> > -            anonymousGallery.mode = "w"
> > -            anonymousGallery.gallery = gallery.gallery
> > -            anonymousGallery.save()
> > -        gallery.delete()
> 
> Petit souci avec ce code. Tel qu'il est écrit, ça signifique que si
> j'ai un user |A| avec les droits d'écritures, et un user |B| avec les
> droits de lectures sur la gallerie, si l'user |A| décide de partir, il
> ne restera plus qu'un user |B| dans la galerie et ce dernier n'aura
> plus les droits d'écriture sur cette gallerie.
> 
> Il faut rajouter external dans la gallerie, s'il n'y a plus personne
> qui a les droits d'écriture sur cette gallerie
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1469/files#r17634166.

firm1 replied:
> Dans les faits, personne n'as accès à "Auteur externe"

Si, l'équipe comm a accès à auteur externe.

> qu'est ce que ça change d'avoir une gallerie en lecture seule ?

Quand l'auteur se désinscrit, le compte auteur externe sert de tampon en attendant que quelqu'un d'autre reprenne le tuto si possible. Donc si personne n'est proprio de la gallerie, le tutoriel ne pourra plus être repris. D'ou le fait de donner à auteur externe les pleins pouvoir sur la gallerie lors de la désinscription.

pierre-24 replied:
Je suis obligé d'être d'accord avec Firm1, du coup.

artragis replied:
> Quand l'auteur se désinscrit, le compte auteur externe sert de tampon
> en attendant que quelqu'un d'autre reprenne le tuto si possible. Donc si
> personne n'est proprio de la gallerie, le tutoriel ne pourra plus être
> repris. D'ou le fait de donner à auteur externe les pleins pouvoir sur
> la gallerie lors de la désinscription.

Je le redis, le passage que tu cites ne parle que des galeries du
**forum** pas celles des tutos qui sont **déjà** traitées **comme il
faut** avant.

Le 17/09/2014 09:51, firm1 a écrit :

> Quand l'auteur se désinscrit, le compte auteur externe sert de tampon
> en attendant que quelqu'un d'autre reprenne le tuto si possible. Donc
> si personne n'est proprio de la gallerie, le tutoriel ne pourra plus
> être repris. D'ou le fait de donner à auteur externe les pleins
> pouvoir sur la gallerie lors de la désinscription.

firm1 replied:
> Je le redis, le passage que tu cites ne parle que des galeries du **forum** pas celles des tutos qui sont **déjà** traitées **comme il faut** avant.

Bah non justement, c'est l'objet de ma remarque. Si tu fais un test en local tu devrais t'en rendre compte.

artragis commented on zds/member/views.py:
```diff
@@ -85,6 +89,89 @@ def index(request):
+            tuto.delete_entity_and_tree()
+        else:
+            if tuto.authors.count() == 1:
+                tuto.authors.add(external)
```
Ok, c'est ici qu'il faut ajouter le fait qu'on met external dans la gallery du tuto, comme ça, le commentaire plus loin n'a plus de sens.

pierre-24 committed:
Fix la documentation sur remarques de Firm1

artragis committed:
Permet de donner la gallery d'un tutoriel non supprimé à external avec droit d'écriture.

artragis committed:
Corrige les tests pour éviter les doublons

Rend les factory plus cohérentes dans les tutoriels

artragis committed:
pep8

SpaceFox committed:
Merge pull request #1517 from artragis/desinscription_fix_gallery

[PR Désinscription] Permet de donner la gallery d'un tutoriel non supprimé à external

Alex-D committed:
Merge pull request #1515 from pierre-24/minifix-doc

[PR désinscription] Fix la documentation sur remarques de Firm1

pierre-24 commented:
Bon, du coup, qu'est ce qui bloque encore, ici ?

SpaceFox commented:
J'allais poser la même question.

@artragis on peut tester ?

artragis commented:
Pour cette réponse, je laisse la parole à Obama :

# "Yes we can"

pierre-24 commented:
... Encore une fois tout le process, ou on peut monter en préprod ?

artragis commented:
je pense qu'avant la preprod, il faut qu'au moins une personne ait testé
le dernier bugfix (passage complet des galeries à externe quand le tuto
est publié seul)

pierre-24 commented:
Je prend :) Par contre, faut que quelqu'un teste absolument que le mot de passe de "Auteur externe" est enfin `external` en local, parce que je peux pas faire ma propre QA.

pierre-24 commented:
Rapport de QA : la galerie devient bien la propriété d'Auteur externe, donc ça c'est bon. Par contre, il n'as pas les droits d'écriture dessus, et je ne sais pas si c'est prévu ou pas (ping @firm1 ?). je me doute que non, mais c'est juste pour vérifier.

firm1 commented:
L'auteur externe (ou quelqu'un) DOIT avoir les droits d'écriture sur la
galerie, sinon personne ne pourra plus y toucher.

pierre-24 commented:
... M'en doutais :/

artragis commented:
C'est prévu mais je vois pas quoi faire de plus que faire ce que j'ai
fait. Que quelqu'un s'en occupe du coup.

pierre-24 commented:
... Je pense avoir trouvé. Près pour une PR _one-byte-change_ ?

artragis commented:
ready
^^
Le 22/09/2014 21:44, Pierre Beaujean a écrit :

> ... Je pense avoir trouvé. Près pour une PR /one-byte-change/ ?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1469#issuecomment-56429410.

SpaceFox commented:
Go go go !

pierre-24 committed:
Fix d'une permission incorrecte

pierre-24 committed:
Ajout de documentation sur les galleries

dralliw committed:
Merge pull request #1521 from pierre-24/fix-permissions

[PR désinscription] Fix d'une permission incorrecte

firm1 commented on zds/member/tests/tests_views.py:
```diff
@@ -75,6 +98,226 @@ def test_register(self):
+        writingArticleAlone = ArticleFactory()
+        writingArticleAlone.authors.add(user.user)
+        writingArticleAlone.save()
+        writingArticle2 = MiniTutorialFactory()
```
`writingArticle2 = MiniTutorialFactory()` => `writingArticle2 = ArticleFactory()`

firm1 commented:
Avant de faire passer cette PR, et vu sa sensibilité, il serait préférable que les Tests Unitaires couvrent tous les cas de désinscription (on ne peut pas se permettre d'avoir de la regression là dessus). 
Actuellement il y'a pas mal de choses qui ne sont pas vérifiés dans les tests comme : 
- lorsqu'on est pas connecté, on a pas accès à la désinscription
- le cas des tutoriels article en cours de validation et/ou en beta n'est pas traité.
- pour un article en rédaction, lorsqu'on se désinscrit, on ne vérifie pas que le tutoriel n'existe plus physiquement
- pour un tuto/article publié, lorsqu'on se désinscrit on ne vérifie pas que le tutoriel existe encore physiquement et est toujours accessible sur son lien public.
- pour les tutoriels, lorsqu'on se désinscrit on ne vérifie pas que les galleries associées sont supprimées (de manière logique et physique) en même temps que le tutoriel et conservées si le tutoriel n'est pas supprimé.
- On ne teste pas le cas ou le membre avait des gallerie simples partagées ou non avec quelqu'un d'autre
- on ne vérifie pas les cas ou le membre qui se désinscrit est en conversation par MP avec 1, 2 ou plusieurs membre. Qu'il soit auteur principal du MP ou pas.

il faudrait aussi prendre en compte la remarque dans le fichier `test_views.py`. Pour moi, si ces tests sont écrit, la QA ne sera qu'une formalité, et on est sur qu'aucune introduction de fonctionnalité ne cassera la désinscription.

pierre-24 commented:
Firm1, je sais que tu as raison et si personne ne s'en occupe d'ici disons ... Ce soir, je m'y collerai. Ceci dit, ce test en question prend UN MAX DE TEMPS à s'exécuter, ce qui devient vite (voir très vite) énervant quand on fait le test juste pour s'assurer qu'on a pas fait de faute de syntaxe ou que la réponse renvoyée par le site est effectivement une 302 et non une 200. Du coup ... Y'a pas un truc pour partiellement accélérer le bouzin ? (genre, par exemple, la validation d'un tutoriel, c'est LONG!).

artragis commented:
J'ai écrit quelques tests qui gèrent une partie des manques. Je les push
dès que celui que je viens d'écrire cesse de bugger (faute de frappe
dans le test) et il manquera encore la moitié à faire environ.
Le 23/09/2014 09:30, Pierre Beaujean a écrit :

> Firm1, je sais que tu as raison et si personne ne s'en occupe d'ici
> disons ... Ce soir, je m'y collerai. Ceci dit, ce test en question
> prend UN MAX DE TEMPS à s'exécuter, ce qui devient vite (voir très
> vite) énervant quand on fait le test juste pour s'assurer qu'on a pas
> fait de faute de syntaxe ou que la réponse renvoyée par le site est
> effectivement une 302 et non une 200. Du coup ... Y'a pas un truc pour
> partiellement accélérer le bouzin ? (genre, par exemple, la validation
> d'un tutoriel, c'est LONG!).
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1469#issuecomment-56484625.

firm1 commented:
> Y'a pas un truc pour partiellement accélérer le bouzin ?

De ce que j'ai vu en regardant un peu le code, la partie qui doit piquer un max de temps c'est au moment du (ou tentative de) téléchargement des images de tutos pendant la publication. Plus précisément [cette variable](https://github.com/zestedesavoir/zds-site/blob/dev/zds/tutorial/factories.py#L15) qui contient tous les cas improbables de chargement d'exception d'images. D'autant plus que si tu as une mauvaise connexion, le test prend encore plus de temps.

Je pense donc que pour accelerer un peu le test, tu peux modifier la factory des tutoriels en donnant la possibilité, lors de la création d'un factory de passer un paramètre booléen (appelons le `lite`) qui va influer sur le contenu du tutoriel crée.

L'influence reviendra donc à dire que si on fait `bigtuto = BigTutorialFactory(lite=True)` le contenu sera une simple chaine sans image. Et donc, rajouter les conditions qui vont bien [ici](https://github.com/zestedesavoir/zds-site/blob/dev/zds/tutorial/factories.py#L56), [ici](https://github.com/zestedesavoir/zds-site/blob/dev/zds/tutorial/factories.py#L59), [ici](https://github.com/zestedesavoir/zds-site/blob/dev/zds/tutorial/factories.py#L99), [ici](https://github.com/zestedesavoir/zds-site/blob/dev/zds/tutorial/factories.py#L102) et [ici](https://github.com/zestedesavoir/zds-site/blob/dev/zds/tutorial/factories.py#L133), bref, partout ou la variable contenu est appelée.

artragis committed:
ajoute des tests

et améliore l'efficacité

artragis committed:
pep8

SpaceFox committed:
Merge pull request #1524 from artragis/desinscription_tests

ajoute des tests

artragis commented:
> on ne vérifie pas les cas ou le membre qui se désinscrit est en conversation par MP avec 1, 2 ou plusieurs membre. Qu'il soit auteur principal du MP ou pas.

Si lignes 293 à 303 et lignes 331-332

artragis committed:
termine les tests

firm1 commented:
> Si lignes 293 à 303 et lignes 331-332

Ce qui est testé, c'est le cas ou le membre qui part était l'auteur (principal) du MP.

Ce qui n'est pas testé, c'est le cas ou le membre qui part était le participant (donc pas auteur principal) du MP.

artragis committed:
pep8

SpaceFox committed:
Merge pull request #1525 from artragis/desinscription_tests

termine les tests

firm1 commented on zds/member/tests/tests_views.py:
```diff
@@ -75,6 +100,271 @@ def test_register(self):
+        writingTutorialAlone = MiniTutorialFactory(light=True)
+        writingTutorialAlone.authors.add(user.user)
+        writingTutorialAlone.save()
+        writingTutorialAloneGallerPath = os.path.join(MEDIA_ROOT, writingTutorialAlone.gallery.slug)
```
En regardant [la nomencalture d'une gallerie](https://github.com/zestedesavoir/zds-site/blob/dev/zds/gallery/models.py#L19), on se rend compte que le morceau de code `writingTutorialAloneGallerPath = os.path.join(MEDIA_ROOT, writingTutorialAlone.gallery.slug)` ne donne pas vraiment le chemin prévu pour la gallerie.

Ce qui rend donc cette partie du test ainsi que [cette ligne](https://github.com/zestedesavoir/zds-site/pull/1469/files#diff-3faca8bf94c4192b3d380f14a9c82270R315) invalide.

Tu dois le remplacer par quelque chose du genre `writingTutorialAloneGallerPath = os.path.join(MEDIA_ROOT, os.path.join('galleries', str(writingTutorialAlone.gallery.pk)))` et vérifier que ce répertoire existe avant désinscription.

artragis replied:
> et vérifier que ce répertoire existe avant désinscription

avant ou après?

`|writingTutorialAloneGallerPath = os.path.join(MEDIA_ROOT,
writingTutorialAlone.gallery.slug)` : très clairement c'est une méthode
à envoyer dans le modèle ça.

|
|

> et vérifier que ce répertoire existe avant désinscription

artragis replied:
en tout cas tu avais raison d'insister, il y avait un bug. Je commit.
Le 23/09/2014 17:21, firm1 a écrit :

> |os.path.join(MEDIA_ROOT, os.path.join('galleries',
> str(writingTutorialAlone.gallery.pk)))|

artragis committed:
fix bug de suppression de gallery dans un tuto privé

et le test unitaire qui va avec

artragis committed:
les images doivent être supprimées de la bdd avant toute chose

désolé de la fausse manip :(

artragis committed:
pep8 + typo

SpaceFox committed:
Merge pull request #1526 from artragis/desinscription_tests

[PR Désinscription] Fix bug de suppression de gallery dans un tuto privé

artragis commented:
Je pense qu'on peut tenter de passer ça sur la preprod non?

SpaceFox commented:
Si c'est bon pour vous, je lance la procédure de release de la v1.1 ce soir.

Ça implique concrètement qu'on envoie **toute la v1.1 en préprod** et qu'on a 2 semaines pour la corriger ou l'annuler.

artragis commented:
yey
Le 24/09/2014 10:10, SpaceFox a écrit :

> Si c'est bon pour vous, je lance la procédure de release de la v1.1 ce
> soir.
> 
> Ça implique concrètement qu'on envoie _toute la v1.1 en préprod_ et
> qu'on a 2 semaines pour la corriger ou l'annuler.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1469#issuecomment-56638002.

pierre-24 commented:
Enfin :)

firm1 commented on zds/member/tests/tests_views.py:
```diff
@@ -75,6 +100,271 @@ def test_register(self):
+        result = self.client.post(
+            reverse('zds.member.views.unregister'),
+            follow=False)
+        self.assertEqual(result.status_code, 302)
```
`self.assertEqual(result.status_code, 302)` : ce test ne vérifie pas l'assertion "lorsqu'on est pas connecté, on a pas accès à la désinscription", la preuve en est que [juste en dessous](https://github.com/zestedesavoir/zds-site/pull/1469/files#diff-3faca8bf94c4192b3d380f14a9c82270R120), tu fais le même test pour l'utilisateur connecté.

Il faut rajouter une vérification que en mode déconnecté, on est _redirigé_ sur la page de login.

firm1 commented:
Un cas d'utilisation que je ne vois pas traité dans la PR : 
- j'écris un tuto tout seul
- je le mets en beta (donc un topic dédié est ouvert pour ce tutoriel)
- je me désinscris
- mes messages seront annonymisées (bien), le tutoriel en question sera supprimé (bien), le topic n'est pas maj pour indiquer que le tutoriel est désormais innaccesible (**pas bien**).

SpaceFox commented:
> je le mets en beta (donc un topic dédié est ouvert pour ce tutoriel)

Sauf erreur de ma part, aujourd'hui le topic n'est pas créé automatiquement (enfin ça ne l'a pas fait quand j'ai mis mon tuto en bêta). Donc on se retrouve avec un cas qui n'est pas gérable automatiquement.

firm1 commented:
> Sauf erreur de ma part, aujourd'hui le topic n'est pas créé automatiquement (enfin ça ne l'a pas fait quand j'ai mis mon tuto en bêta). Donc on se retrouve avec un cas qui n'est pas gérable automatiquement.

[Si](https://github.com/zestedesavoir/zds-site/pull/1415)

SpaceFox commented:
Dans les derniers retours, est-ce qu'il y a quoi que ce soit qui bloque _encore_ le lancement de la release de la v1.1 ?

artragis committed:
prévient les utilisateurs qu'un tuto en béta a été supprimé en cas de désinscription

artragis commented:
l'avertissement du topic pour la béta test. Je le fais en ce moment
même. La PR arrive rapidement.

Le 24/09/2014 16:27, SpaceFox a écrit :

> Dans les derniers retours, est-ce qu'il y a quoi que ce soit qui
> bloque /encore/ le lancement de la release de la v1.1 ?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1469#issuecomment-56677683.

SpaceFox commented:
Ok

2014-09-24 16:35 GMT+02:00 artragis notifications@github.com:

> l'avertissement du topic pour la béta test. Je le fais en ce moment
> même. La PR arrive rapidement.
> 
> Le 24/09/2014 16:27, SpaceFox a écrit :
> 
> > Dans les derniers retours, est-ce qu'il y a quoi que ce soit qui
> > bloque /encore/ le lancement de la release de la v1.1 ?
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > <
> > https://github.com/zestedesavoir/zds-site/pull/1469#issuecomment-56677683
> > .
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1469#issuecomment-56679004
> .

firm1 commented:
> Dans les derniers retours, est-ce qu'il y a quoi que ce soit qui bloque encore le lancement de la release de la v1.1 ?

je ne pense pas. Il faut juste bien couvrir les tests.

Ceci dit, au lancement de la v1.1, il faut pensez à faire un snapshot la preprod **AVANT** de pousser la V1.1 dessus.

SpaceFox commented:
Ping @geoffreyc y'a que toi qui peut faire ça !

SpaceFox committed:
Merge pull request #1530 from artragis/desinscription_tests

prévient les utilisateurs qu'un tuto en béta a été supprimé

Alex-D commented:
Je rêve ou c'est mergé et on peut faire la fête ? :D

SpaceFox commented:
Tu ne rêves pas. Mais il va falloir que je mette en PRP avant de faire la fête.
